test/contrib/s3_test.py

```diff
@@ -503,6 +503,11 @@ def test_copy_dir(self):
             copy_size = s3_client.get_key(s3_dest + str(i)).size
             self.assertEqual(original_size, copy_size)
 
+    def test__path_to_bucket_and_key(self):
+        self.assertEqual([('bucket', 'noquestionmarkhere'), ('bucket', 'questionmark?here')],
+                         [S3Client._path_to_bucket_and_key('s3://bucket/noquestionmarkhere'),
+                          S3Client._path_to_bucket_and_key('s3://bucket/questionmark?here')])
```

Comment:
While this does what you want it to, i believe it's generally better to split unittest tests into very small tests or just having a couple quick/easy assertions. Such as

```python
def test__path_to_bucket_and_key(self):
    self.assertEqual(('bucket', 'dir'), S3Client._path_to_bucket_and_key('s3://bucket/dir))

def test__path_to_bucket_and_key_with_question_mark(self):
    self.assertEqual(('bucket', 'dir?arg'), S3Client._path_to_bucket_and_key('s3://bucket/dir?arg))
```

or even within 1 unittest test, with multiple assertions:

```python
def test__path_to_bucket_and_key(self):
    self.assertEqual(('bucket', 'dir'), S3Client._path_to_bucket_and_key('s3://bucket/dir))
    self.assertEqual(('bucket', 'dir?arg'), S3Client._path_to_bucket_and_key('s3://bucket/dir?arg))
```

The first example above would present easier identification during a test failure (bc the test named `test__path_to_bucket_and_key_with_question_mark` would fail and that's very clear what the problem is). The latter may be easier to read, but test failure wouldn't be as explicit.

Reply:
@dlstadther Seems fair! I went with the first approach, what do you think? :)